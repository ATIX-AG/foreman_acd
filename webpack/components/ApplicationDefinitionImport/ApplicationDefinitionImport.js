import React from 'react';
import PropTypes from 'prop-types';
import {
  Icon,
  Button,
  MessageDialog,
  Table,
  inlineEditFormatterFactory,
} from 'patternfly-react';
import * as resolve from 'table-resolver';
import { translate as __ } from 'foremanReact/common/I18n';
import Select from 'foremanReact/components/common/forms/Select';
import CommonForm from 'foremanReact/components/common/forms/CommonForm';
import RailsData from '../common/RailsData';
import { EasyHeaderFormatter } from '../../helper';

class ApplicationDefinitionImport extends React.Component {
  onFileChange = event => {
    // Update the state
    this.setState({ selectedFile: event.target.files[0] });
  };

  static isEditing({ rowData }) {
    return rowData.backup !== undefined;
  }

  componentDidMount() {
    const { ansiblePlaybookServices, hostgroups } = this.props;

    const {
      initApplicationDefinitionImport,
      changeEditApplicationDefinitionImportService,
    } = this.props;

    this.headerFormatter = EasyHeaderFormatter;

    const inlineEditFormatterImpl = {
      renderValue: (value, additionalData) => (
        <td>
          <span className="static">{value}</span>
        </td>
      ),
      renderEditSelect: (value, additionalData, options) => (
        <td className="editing">
          <Select
            value={value.toString()}
            onChange={e =>
              changeEditApplicationDefinitionImportService(
                e.target.value,
                additionalData
              )
            }
            options={options}
            allowClear
            key="key"
          />
        </td>
      ),
    };

    const inlineEditFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => this.constructor.isEditing(additionalData),
      renderValue: (value, additionalData) => {
        let prettyValue = value;
        if (additionalData.property === 'hostgroup') {
          prettyValue = hostgroups[value];
        }
        return inlineEditFormatterImpl.renderValue(prettyValue, additionalData);
      },
      renderEdit: (value, additionalData) => {
        if (additionalData.property === 'hostgroup') {
          if (additionalData.rowData) {
            return inlineEditFormatterImpl.renderEditSelect(
              value,
              additionalData,
              hostgroups
            );
          }
          return inlineEditFormatterImpl.renderValue(
            hostgroups[value],
            additionalData
          );
        }
        return inlineEditFormatterImpl.renderValue(value, additionalData);
      },
    });
    this.inlineEditFormatter = inlineEditFormatter;

    initApplicationDefinitionImport(
      ansiblePlaybookServices,
      this.headerFormatter,
      this.inlineEditFormatter
    );
  }

  render() {
    const {
      showAlertModal,
      alertModalText,
      alertModalTitle,
      closeAlertModal,
      ansiblePlaybookServices,
      columns,
      handleImportAnsiblePlaybook,
    } = this.props;

    return (
      <span>
        <MessageDialog
          show={showAlertModal}
          onHide={closeAlertModal}
          primaryAction={closeAlertModal}
          primaryActionButtonContent={__('OK')}
          primaryActionButtonBsStyle="danger"
          icon={<Icon type="pf" name="error-circle-o" />}
          title={alertModalTitle}
          primaryContent={alertModalText}
        />
        <div>
          <CommonForm
            label="Application Definition file"
            className="custom-file-upload"
          >
            <input type="file" onChange={this.onFileChange} />
          </CommonForm>
          <CommonForm>
            <Button
              bsStyle="default"
              onClick={e =>
                handleImportAnsiblePlaybook(this.state.selectedFile, e)
              }
            >
              {__('Import')}{' '}
            </Button>
          </CommonForm>
        </div>

        {ansiblePlaybookServices.length > 0 ? (
          <div className="form-group">
            <Table.PfProvider
              striped
              bordered
              hover
              dataTable
              columns={columns}
              components={{
                body: {
                  cell: cellProps => cellProps.children,
                },
              }}
            >
              <Table.Header headerRows={resolve.headerRows({ columns })} />
              <Table.Body
                rows={ansiblePlaybookServices}
                rowKey="id"
                onRow={(rowData, { rowIndex }) => ({
                  role: 'row',
                  isEditable: () => this.isEditing({ rowData }),
                  last: ansiblePlaybookServices.length - 1,
                })}
              />
            </Table.PfProvider>
          </div>
        ) : null}
        <RailsData
          key="applications_definition_import"
          view="app_definition_import"
          parameter="services"
          value={JSON.stringify(ansiblePlaybookServices)}
        />
      </span>
    );
  }
}

ApplicationDefinitionImport.defaultProps = {
  // error: {},
  showAlertModal: false,
  alertModalText: '',
  alertModalTitle: '',
  columns: [],
  ansiblePlaybookServices: [],
};

ApplicationDefinitionImport.propTypes = {
  hostgroups: PropTypes.object.isRequired,
  // error: PropTypes.object,
  ansiblePlaybookServices: PropTypes.array,
  // editMode: PropTypes.bool.isRequired,
  columns: PropTypes.array,
  showAlertModal: PropTypes.bool,
  alertModalText: PropTypes.string,
  alertModalTitle: PropTypes.string,
  changeEditApplicationDefinitionImportService: PropTypes.func.isRequired,
  closeAlertModal: PropTypes.func.isRequired,
  handleImportAnsiblePlaybook: PropTypes.func.isRequired,
  initApplicationDefinitionImport: PropTypes.func.isRequired,
};

export default ApplicationDefinitionImport;
