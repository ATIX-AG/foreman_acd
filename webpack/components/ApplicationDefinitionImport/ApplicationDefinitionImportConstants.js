export const APPLICATION_DEFINITION_IMPORT_INIT =
  'INIT_APPLICATION_DEFINITION_IMPORT_INIT';
export const APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE =
  'APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE';
export const APPLICATION_DEFINITION_IMPORT_FILE_FAILURE =
  'APPLICATION_DEFINITION_IMPORT_FILE_FAILURE';
export const APPLICATION_DEFINITION_IMPORT_FILE_REQUEST =
  'APPLICATION_DEFINITION_IMPORT_FILE_REQUEST';
export const APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS =
  'APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS';
export const APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL =
  'APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL';
