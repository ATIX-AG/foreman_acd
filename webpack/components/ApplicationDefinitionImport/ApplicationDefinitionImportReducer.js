import Immutable from 'seamless-immutable';

import { cloneDeep, findIndex, findLastIndex } from 'lodash';

import { translate as __ } from 'foremanReact/common/I18n';

import {
  APPLICATION_DEFINITION_IMPORT_INIT,
  APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE,
  APPLICATION_DEFINITION_IMPORT_FILE_FAILURE,
  APPLICATION_DEFINITION_IMPORT_FILE_REQUEST,
  APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS,
  APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL,
} from './ApplicationDefinitionImportConstants';

export const initialState = Immutable({
  name: false,
  error: { errorMsg: '', status: '', statusText: '' },
});

const applicationDefinitionImportConf = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case APPLICATION_DEFINITION_IMPORT_INIT: {
      return state.merge(payload);
    }

    case APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL: {
      return state.merge({
        showAlertModal: false,
        alertModalTitle: '',
        alertModalText: '',
      });
    }

    case APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE: {
      const services = cloneDeep(state.ansiblePlaybookServices);
      const index = findIndex(services, { id: payload.rowData.id });

      services[index][payload.property] = payload.value;

      return state.set('ansiblePlaybookServices', services);
    }

    case APPLICATION_DEFINITION_IMPORT_FILE_FAILURE: {
      return state.merge({ error: payload.error, loading: false });
    }
    case APPLICATION_DEFINITION_IMPORT_FILE_REQUEST: {
      return state.set('loading', true);
    }
    case APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS: {
      let newState = {};
      const ansibleServices = [];
      const ansiblePlaybookServices =
        payload.ansiblePlaybookServices.ansible_services;

      for (const ind in ansiblePlaybookServices) {
        const value = ansiblePlaybookServices[ind];
        const newRow = { id: value.id, name: value.value, hostgroup: '' };
        newRow.backup = cloneDeep(newRow);
        ansibleServices.push(newRow);
      }

      newState = {
        ansiblePlaybookServices: ansibleServices,
      };

      return state.merge(newState);
    }
    default:
      return state;
  }
};

export default applicationDefinitionImportConf;
