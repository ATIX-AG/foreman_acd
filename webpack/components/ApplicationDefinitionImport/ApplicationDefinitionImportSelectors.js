const applicationDefinitionImportConf = state =>
  state.foremanAcd.applicationDefinitionImportConf;

export const selectEditMode = state =>
  applicationDefinitionImportConf(state).editMode;
export const selectColumns = state =>
  applicationDefinitionImportConf(state).columns;
export const selectAnsiblePlaybookServices = state =>
  applicationDefinitionImportConf(state).ansiblePlaybookServices;
export const selectShowAlertModal = state =>
  applicationDefinitionImportConf(state).showAlertModal;
export const selectAlertModalText = state =>
  applicationDefinitionImportConf(state).alertModalText;
export const selectAlertModalTitle = state =>
  applicationDefinitionImportConf(state).alertModalTitle;
