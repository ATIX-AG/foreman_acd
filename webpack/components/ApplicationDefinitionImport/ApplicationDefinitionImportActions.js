import React from 'react';
import { API, actionTypeGenerator } from 'foremanReact/redux/API';
import { addToast } from 'foremanReact/components/ToastsList';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';


import {
  setModalOpen,
  setModalClosed,
} from 'foremanReact/components/ForemanModal/ForemanModalActions';

import {
  actionHeaderCellFormatter,
} from 'patternfly-react';

import {
  propsToSnakeCase,
  propsToCamelCase,
} from 'foremanReact/common/helpers';

import {
  APPLICATION_DEFINITION_IMPORT_INIT,
  APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL,
  APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE,
  APPLICATION_DEFINITION_IMPORT_FILE_REQUEST,
  APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS,
  APPLICATION_DEFINITION_IMPORT_FILE_FAILURE,
} from './ApplicationDefinitionImportConstants';


export const initApplicationDefinitionImport = (
  ansiblePlaybookServices,
  headerFormatter,
  inlineEditFormatter,
) => dispatch => {
  const initialState = {};

  initialState.columns = [
    {
      property: 'name',
      header: {
        label: 'Service name',
        formatters: [headerFormatter],
        props: {
          index: 0,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        props: {
              index: 0,
            },
        formatters: [inlineEditFormatter]
      }
    },
    {
      property: 'hostgroup',
      header: {
        label: 'Hostgroup',
        formatters: [headerFormatter],
        props: {
          index: 1,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        props: {
              index: 1,
            },
        formatters: [inlineEditFormatter]
      }
    }
  ];

  initialState.ansiblePlaybookServices = ansiblePlaybookServices;

  dispatch({
    type: APPLICATION_DEFINITION_IMPORT_INIT,
    payload: initialState,
  });
};


const errorHandler = (msg, err) => {
  const error = {
    errorMsg: __('Failed to fetch data from server.'),
    statusText: err,
  };
  return { type: msg, payload: { error } };
};

export const closeAlertModal = () => ({
  type: APPLICATION_DEFINITION_IMPORT_CLOSE_ALERT_MODAL,
  payload: {}
});

export const changeEditApplicationDefinitionImportService = (value, additionalData) => ({
  type: APPLICATION_DEFINITION_IMPORT_SERVICE_EDIT_CHANGE,
  payload: {
    value,
    ...additionalData,
  },
});


export const handleImportAnsiblePlaybook = (
  file, e,) => async dispatch => {
  e.preventDefault();
  dispatch({
    type: APPLICATION_DEFINITION_IMPORT_FILE_REQUEST,
    payload: {
      app_definition_file: file,
    }});
  try {
    const formData = new FormData();

    // Update the formData object
    formData.append(
      "app_definition_file",
      file,
      file.name
    );
    const { data } = await API.post(
      '/acd/app_definitions/handle_file_upload', formData);
    dispatch(
        addToast({
          type: 'success',
          message: sprintf(
            __('Sucessfully synced imported app template')
          ),
          key: APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS,
        })
      );
      return dispatch({
        type: APPLICATION_DEFINITION_IMPORT_FILE_SUCCESS,
        payload: {
          ansiblePlaybookServices: data,
        },
        response: data,
      });
    } catch (error) {
      dispatch(
        addToast({
          type: 'error',
          message: sprintf(__('Error occurred while importing app template: %s'), error.response.data.message),
          key: APPLICATION_DEFINITION_IMPORT_FILE_FAILURE,
        })
      );
      return dispatch({
        type: APPLICATION_DEFINITION_IMPORT_FILE_FAILURE,
        payload: {
          error: error,
        },
        response: error,
      });
    }
};
