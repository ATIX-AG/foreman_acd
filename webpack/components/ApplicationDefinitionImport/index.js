import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './ApplicationDefinitionImport.scss';
import ApplicationDefinitionImport from './ApplicationDefinitionImport';
import * as ApplicationDefinitionImportActions from './ApplicationDefinitionImportActions';

import {
  selectEditMode,
  selectColumns,
  selectAnsiblePlaybookServices,
  selectShowAlertModal,
  selectAlertModalText,
  selectAlertModalTitle,
} from './ApplicationDefinitionImportSelectors';

const mapStateToProps = state => ({
  editMode: selectEditMode(state),
  columns: selectColumns(state),
  showAlertModal: selectShowAlertModal(state),
  alertModalText: selectAlertModalText(state),
  alertModalTitle: selectAlertModalTitle(state),
  ansiblePlaybookServices: selectAnsiblePlaybookServices(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(ApplicationDefinitionImportActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ApplicationDefinitionImport);
