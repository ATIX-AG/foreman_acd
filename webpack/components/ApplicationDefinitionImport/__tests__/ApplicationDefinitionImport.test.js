import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ApplicationDefinitionImport from '../ApplicationDefinitionImport';

const noop = () => {};

const fixtures = {
  'should render application definition import': {
      location: "Default Location",
      organization: "Default Organization",
      hostgroups: [],
      ansiblePlaybookServices: [],
    initApplicationDefinitionImport: noop,
    addApplicationDefinitionImportService: noop,
    closeAlertModal: noop,
  },
};

describe('ApplicationDefinitionImport', () =>
  testComponentSnapshotsWithFixtures(ApplicationDefinitionImport, fixtures));
