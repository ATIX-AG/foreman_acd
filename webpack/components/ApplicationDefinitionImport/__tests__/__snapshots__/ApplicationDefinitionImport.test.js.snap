// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationDefinitionImport should render application definition import 1`] = `
<span>
  <Component />
  <Component />
  <MessageDialog
    accessibleDescription=""
    accessibleName=""
    className=""
    enforceFocus={true}
    footer={null}
    icon={
      <Icon
        name="error-circle-o"
        type="pf"
      />
    }
    onHide={[Function]}
    primaryAction={[Function]}
    primaryActionButtonBsStyle="danger"
    primaryActionButtonContent="OK"
    primaryContent=""
    secondaryAction={[Function]}
    secondaryActionButtonBsStyle="default"
    secondaryActionButtonContent={null}
    secondaryContent={null}
    show={false}
    title=""
  />
  <div>
    <CommonForm
      className="custom-file-upload"
      label="Application Definition file"
    >
      <input
        onChange={[Function]}
        type="file"
      />
    </CommonForm>
    <CommonForm>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        onClick={[Function]}
      >
        Import
         
      </Button>
    </CommonForm>
  </div>
  <RailsData
    key="applications_definition_import"
    parameter="services"
    value="[]"
    view="app_definition_import"
  />
</span>
`;
